src/main/java/org/broadinstitute/dsm/route/KitUploadRoute.java

```diff
@@ -161,6 +163,13 @@ public Object processRequest(Request request, Response response, String userId)
                         logger.info("placing order with external shipper");
                         ExternalShipper shipper = (ExternalShipper) Class.forName(DSMServer.getClassName(kitRequestSettings.getExternalShipper())).newInstance();
                         shipper.orderKitRequests(orderKits, easyPostUtil, kitRequestSettings, shippingCarrier);
+                        // mark kits as transmitted so that background jobs don't try to double order it
```

Comment:
could it be that the background job would pick them up before we set them to transmitted? maybe the method to write them into the ddp_kit_request table should already set them to transmitted?

Reply:
Yes, this is a possibility.  `GBFOrderFinder` does put a read lock onto ddp_kit_request, but that only protects against multiple instances of the eventual CF.  What we'd need to do here is to put read locks on the orders to be issued during the upload as well.  Would involve doing a re-read on `ddp_kit_request`s in the list here with a read lock, and then double checking that `order_transmission_date` is still `null` in order to place the order.  How's that sound?

Reply:
I think easiest way would be to just add a transmitted date when adding kit request to the ddp_kit_request table when uploading

Reply:
In that case, there will still be a gap where a separate transaction could grab the orders.  How about if we wrap the whole thing in a separate transaction?  One call to GBF, so either the whole list goes or it fails.  Something like:

```
TransactionWrapper.inTransaction(conn -> {
                    uploadKit(ddpInstance, kitType, kitUploadObjects, kitHasSubKits, kitRequestSettings, easyPostUtil, userIdRequest, kitTypeName,
                            uploadAnyway, invalidAddressList, duplicateKitList, orderKits, specialKitList, upload, kitUploadReason, shippingCarrier, conn);

                    //only order if external shipper name is set for that kit request
                    if (StringUtils.isNotBlank(kitRequestSettings.getExternalShipper())) {
                        try {
                            logger.info("placing order with external shipper");
                            ExternalShipper shipper = (ExternalShipper) Class.forName(DSMServer.getClassName(kitRequestSettings.getExternalShipper())).newInstance();
                            shipper.orderKitRequests(orderKits, easyPostUtil, kitRequestSettings, shippingCarrier);
                            // mark kits as transmitted so that background jobs don't try to double order it
                            for (KitRequest orderKit : orderKits) {
                                KitRequestShipping.markOrderTransmittedAt(conn, orderKit.getExternalOrderStatus(), Instant.now());
                            }
                        }
                        catch (RuntimeException e) {
                            logger.error("Failed to sent kit request order to " + kitRequestSettings.getExternalShipper(), e);
                            response.status(500);
                            return new Result(500, "Failed to sent kit request order to " + kitRequestSettings.getExternalShipper());
                        }
                    }
                });
 ```